### DISCUSSION TITLE
How to create files based on items in a list?

### DISCUSSION THREAD

**Discussion by cdhesse** [copier-org/copier#1240] [Q&A] [answered]
I have a list of environments, such as dev, qa, prod.  I want to create <environment>.properties files dynamically in a single directory.  Given the input like this...

environments:
  - dev
  - qa
  - prod

I need

./conf
    dev.properties
    qa.properteis
    prod.properties

The list of environments can be any value, so I don't know the names of the environments before template generation.

Another viable input could be:

environments:
  - prod
  - nonprod

or

environments:
  - mysandbox
  - yoursandbox
  - bobsdev
  - dev
  - qa
  - prod
  - etc

Is there a way to do this?

**Comment by sisp** [comment, marked as answer]
I'm afraid this is currently not possible (in a good way). It's related to https://github.com/copier-org/copier/discussions/908 and I proposed a feature design in https://github.com/copier-org/copier/discussions/908#discussioncomment-5466993 but haven't had the time to work on it. But I think it would do what you need.

You might try using [tasks](https://copier.readthedocs.io/en/stable/configuring/#tasks), but you need to be a bit careful when _updating_ a project to not overwrite those files again, and (if ever needed) deleting previously generated files (e.g. when a list item gets removed) isn't automatically handled.

**Reply by sisp** [reply]
Also somewhat related to #1236, I believe.

**Reply by cdhesse** [reply]
I believe[ #908 (comment)](https://github.com/orgs/copier-org/discussions/908#discussioncomment-5466993) would indeed be exactly what i need, and a powerful feature.  Coming from cookiecutter (which i found first and moved to copier based on limitations there) - i was able to do this use case with hooks.  You have my upvote for your solution fwiw! 

Thank you for the prompt answer!  Awesome work as well!

**Reply by sisp** [reply]
Copier tasks are essentially equivalent to the Cookiecutter post-gen hook when you run a Python script as a task. But tasks are also run as part of the update algorithm, which Cookiecutter doesn't have, so you need to be a bit careful.
